addons/mail/static/tests/helpers/mock_server.js

```diff
@@ -861,6 +862,24 @@ MockServer.include({
                 this._widget.call('bus_service', 'trigger', 'notification', [notifConfirmUnpin]);
             }
             return;
+        } else if (commandName === 'who') {
+            for (const channel of channels) {
+                const members = channel.members.map(memberId => this._getRecords('res.partner', [['id', '=', memberId]])[0].name);
+                let message = "You are alone in this channel.";
+                if (members.length > 0) {
+                    message = `Users in this channel: ${members.join(', ')} and you`
+                }
+                const notification = [
+                    ["dbName", 'res.partner', this.currentPartnerId],
+                    {
+                        'body': `<span class="o_mail_notification">${message}</span>`,
+                        'channel_ids': [channel.id],
+                        'info': 'transient_message',
+                    }
+                ];
+                this._widget.call('bus_service', 'trigger', 'notification', [notification]);
+            }
+            return;
```

Comment:
I think @jgi-odoo has a PR which introduced `mail.channel_command` mock server data. Would make sense that the mocked behaviour of executing command relies on that mocked model & data.

It's best behaviour of `/who` being dictate in test that use it, rather than generalizing to all tests.

Reply:
Not when "who" is globally defined on the server. This is just asking for copy/paste in every test. We don't want to have to handle those separately if/when we get rid of the mock server, it should be as transparent as possible. (so global on server = global on mock)

Reply:
All data in test should be transparent.The behaviour of `/who` being fully relying on current way mock server is implemented is very very bad.

Using the actual server is less bad, but still doesn't rule in favour of server code being actually good.
The tests are much better written with explicit `/who` data. No hidden data. If we later use actual server and don't change its code, I much prefer deleting this part of test than actually adapt it: the effort is much less.

Please don't blindly copy bad stuff from server in mocked server, especially when we advocate for high quality tests which I'm all in favour of.

Reply:
This is the only valid implementation of `who`, it is the one returning the list of channel members... It's in mock server for the same reasons all other rpc are in the mock server. We don't want to return to mocked rpcs in every test, and even less mocked rpc for commands.

Reply:
I feel channel commands are data that contain behaviour. The actual python code is ugly because it makes everything behavioural by hiding data in implementation (method whose part of name is the command itself).

IMO it should have been the inverse: data which part of it contains the behaviour. This approach is much more modular, exactly what the command should have been in the first place.

Reply:
It's just the way we handle them at init messaging that looks like data. I think they need to be seen for what they really are: real python method/rpc just like any other channel interaction.

And it's just that instead of clicking a UI element, they triggered by faking sending a message, which is another reason why we consider them strangely.

Honestly we could just as well remove them from init and hardcode them in JS, it would be the same.

Reply:
Hardcode in JS sure no. No modeling of `mail.channel_command`s in JS models, this would lead to list of commands in channel info payload or over reliance of python with RPC everywhere (to the point data of RPC directly used in template... we chose not to do that because consistency and speed are valuable, therefore client should be much more knowledgeable than it previously was, hence our client-side ORM was born).

Python didn't make data for channel commands. I don't know why not, but I suspect this question has just not been asked and the current implementation is the one that was picked certainly because it was very fast to make with the functional intended behaviour. I'm pointing out this solution is flawed, and it becomes apparent when we care for quality of code (more than just presence of feature).

Our aim is to make the best code we can do. We have to accept good stuff usually takes some time, and time is also important, so we have to deploy _not-quite-nice-as-expected_ code. Here there's no such motivation that we need to rush, and what you suggest @seb-odoo is the _not-quite-nice-as-expected_ code we should do right now: it takes relatively the same time to do the nicer code than this one.

I insist: this is not the right direction, and it's not worth going in the wrong direction.

Reply:
Commands are just "alias" from "fake message" to "do some action". This is purely client side, there's no reason the server should know or return them. And particularly not from channel info, not sure where you got that idea from my message.

JS should define allowed commands and it should know how to parse them to transform them into proper action. Action can be some JS code, or it could be calling server. At this point this is just like any other UI interaction, of which the trigger was a fake message instead of a button click. Command should just be transform fake message to execution of code. It's just as hard-coded as displaying a button on the UI and binding it's event is hardcoded.

When the action can be done in JS then sure, but when the action relies on the server its result must be mocked. Currently all commands rely on server, so all of them (that we want to test) should be mocked like it is here. Here we are just doing a fix in stable and we happen to need to make an rpc + a bus notification from the server in the tested flow -> mock server.

We can discuss this further for what we want to do with commands in master.

Reply:
Yes, the nature of PR is fix, so more incline to just fix fast. Master is clearly a better place to think of that.
The PR of @jgi-odoo that introduces mocked data `mail.channel_command` is also a fix in 14.0, so unfortunately both PRs while fixes introduce a big inconsistency of `mail.channel_commands` being both data and behaviour. It doesn't help both tests use `/who`...

I'm a bit annoyed of this inconsistency... It's a bit sad if we have to move one step back and go in the direction of this PR (so in @jgi-odoo's PR, make the `/who` global like here)... But have consistency for the moment. Let's decide which approach we take for 14.0, but also let's do our best to not keeping frustrating code for too long (it's quite contagious...).

Reply:
Still no agreement about this. Important because:
- if we do what I say, change requested in this PR.
- if we do what @seb-odoo says, change requested in @jgi-odoo's PR to resolve consistency, and we must do something about acknowledging the weak implementation of mocked `/who` command in some ways (in code, in task, etc.)

Reply:
The RPC of the command must be mocked as it is done here, and as we already did for leave too.

Feel free to create a task explaining what you want to improve in master, because I'm still not sure exactly what is your goal.

Reply:
Yeah, we did that for `/leave` but the problematic design of channel commands didn't make me twitched until now, so `/leave` follows the same issue.

Reply:
A bit concerned with general problem of data/behavior of channel commands, because this is an actual time bomb.
Ok to give a pass for this PR...